### PR TITLE
web: optimize log rendering

### PR DESCRIPTION
**Pull request by nicks**
Hello @hyu, @lizzthabet,

Please review the following commits I made in branch nicks/react-optimize2:

c6960421c234c77d200da5ab53df259653560d26 (2021-06-17 15:31:31 -0400)
web: optimize log rendering

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
I'm hoping this will fix some of the complaints in the user research about the web ui using a lot of CPU when there are lots of logs.

setState(null) will ensure the app doesn't re-render when the only change that came in was new logs

File: web/src/HUD.tsx [hunk 1]
```diff
@@ -413,5 +413,26 @@ export function mergeAppUpdate<K extends keyof HudState>(
...
+  }
+
   return result
 }
+
+function hasChange(result: any, prevState: any): boolean {
+  for (let k in result) {
+    let resultV = result[k] as any
+    let prevV = prevState[k] as any
+    if (resultV != prevV) {
```
Comment (lizzthabet) [inline web/src/HUD.tsx:433]: Does strict equality make a difference here? (Just curious.)

**Comment by lizzthabet** [comment]
That makes sense! Is there any way you'd recommend testing this out locally?

**Comment by nicks** [comment]
The best way to test it manually is to 
1) Install the react profiler (it's a browser extension)
2) run an app with a lot of logs
3) verify that the profiler doesn't see any react re-rendering when the logs come in

File: web/src/HUD.tsx (hunk 1 above)
Comment (nicks) [reply web/src/HUD.tsx]: i don't think it will make a difference, but it certainly doesn't hurt

**Review by lizzthabet — APPROVED** [review]
Sweet! I'm not sure I see any difference in the profiling, but I think I need to read up on how to understand the profiler output better. 😅

**Pull request merged**